Add the builder.input() method steps

zolkis opened the pull request:
Add the input() method algorithm.
Depends on #337


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/zolkis/webnn/pull/364.html" title="Last updated on May 15, 2023, 6:32 AM UTC (25f8982)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/webmachinelearning/webnn/364/a115073...zolkis:25f8982.html" title="Last updated on May 15, 2023, 6:32 AM UTC (25f8982)">Diff</a>

zolkis commented:
@huningxin PTAL :)

huningxin commented on index.bs:
```diff
@@ -1149,9 +1149,35 @@ The [=new=] {{MLGraphBuilder}} constructor steps are:
+        1. If |name| is `undefined` or an empty [=string=], then throw a "{{TypeError}}" {{DOMException}} and stop.
+    1. Let |descriptor| be the second argument.
+        1. If |descriptor| is not an an object that [=implements=] {{MLOperandDescriptor}}, then throw a "{{TypeError}}" {{DOMException}} and stop.
+        1.  If the [=byte length=] of |descriptor| is not supported by the underlying platform, then throw a "{{DataError}}" {{DOMException}} and stop.
```
nit: one extra space before "If"

huningxin commented on index.bs:
```diff
@@ -1149,9 +1149,35 @@ The [=new=] {{MLGraphBuilder}} constructor steps are:
+    1. Let |descriptor| be the second argument.
+        1. If |descriptor| is not an an object that [=implements=] {{MLOperandDescriptor}}, then throw a "{{TypeError}}" {{DOMException}} and stop.
+        1.  If the [=byte length=] of |descriptor| is not supported by the underlying platform, then throw a "{{DataError}}" {{DOMException}} and stop.
+        1. If the [=check dimensions=] steps given |descriptor|.{{MLOperandDescriptor/type}} and |descriptor|.{{MLOperandDescriptor/dimensions}} return `false`, throw a "{{DataError}}" {{DOMException}} and stop.
```
What would happen if user code passes a descriptor for scalar operand (`dimensions` is undefined)? This `input()` algorithm seems only to support tensor operand now.

huningxin replied:
Actually, the `constant(descriptor, bufferView)` algorithm in PR #365 should also address this issue.

zolkis replied:
Right. So this is where conventions about `descriptor` become important.
I am thinking if it's enough to distinguish between vector and scalar solely on the existence of a `dimensions` member. But for now let's continue with that working assumption and keep it in mind.

About `constant(descriptor, bufferView)`: the `descriptor.dimensions` must be defined, otherwise it's an error. There is a polymorphic variant for specifying scalar constant. Should we change that?

zolkis replied:
Done, please check if correct. Using an assert seemed better choice than just making a note.
Forgot to fix the white space error, such a small change that I amended the same commit and force-pushed.

huningxin replied:
> About constant(descriptor, bufferView): the descriptor.dimensions must be defined, otherwise it's an error.

That's fine. No change needed.

huningxin commented on index.bs:
```diff
@@ -1168,8 +1168,10 @@ Create a named {{MLOperand}} based on a descriptor, that can be used as an input
         1. If |descriptor| is not an an object that [=implements=] {{MLOperandDescriptor}}, then throw a "{{TypeError}}" {{DOMException}} and stop.
-        1.  If the [=byte length=] of |descriptor| is not supported by the underlying platform, then throw a "{{DataError}}" {{DOMException}} and stop.
-        1. If the [=check dimensions=] steps given |descriptor|.{{MLOperandDescriptor/type}} and |descriptor|.{{MLOperandDescriptor/dimensions}} return `false`, throw a "{{DataError}}" {{DOMException}} and stop.
+        1. If the [=byte length=] of |descriptor| is not supported by the underlying platform, then throw a "{{DataError}}" {{DOMException}} and stop.
```
"byte length" algorithm also need to handle scalar operand descriptor. Its current definition would break if `descriptor.dimensions` is undefined.

huningxin commented on index.bs:
```diff
@@ -1168,8 +1168,10 @@ Create a named {{MLOperand}} based on a descriptor, that can be used as an input
-        1.  If the [=byte length=] of |descriptor| is not supported by the underlying platform, then throw a "{{DataError}}" {{DOMException}} and stop.
-        1. If the [=check dimensions=] steps given |descriptor|.{{MLOperandDescriptor/type}} and |descriptor|.{{MLOperandDescriptor/dimensions}} return `false`, throw a "{{DataError}}" {{DOMException}} and stop.
+        1. If the [=byte length=] of |descriptor| is not supported by the underlying platform, then throw a "{{DataError}}" {{DOMException}} and stop.
+        1. [=Assert=]: If |descriptor|.{{MLOperandDescriptor/dimensions}} does not [=map/exist=], then |descriptor| defines a scalar input.
```
Move this assertion before byte length check?

zolkis replied:
Actually I moved the byte length check to the vector section.
If we need to check the scalar, that would be an additional step.

huningxin replied:
Sounds good.

huningxin commented:
@zolkis , you may also want to solve the conflicts, squash the commits and edit the commit log before merging. Thanks!

zolkis commented:
@huningxin 
I rebased, squashed, fixed conflicts, verified with build, then force pushed. 
Since there were a lot of other commits in between filing the PR and now, those commits appear here now.
Needed to add one more commit to merge with the main branch.
